File: layer/src/protocol.cpp
#include "maquestlink/protocol.hpp"

#include <bit>
#include <limits>
#include <string>
#include <type_traits>

namespace maquestlink::protocol {
namespace {

template <typename T, bool = std::is_enum_v<T>>
struct WireRawType {
  using type = T;
};

template <typename T>
struct WireRawType<T, true> {
  using type = std::underlying_type_t<T>;
};

template <typename T>
using WireRawTypeT = typename WireRawType<T>::type;

class Writer {
 public:
  template <typename T>
    requires(std::is_integral_v<T> || std::is_enum_v<T> || std::is_floating_point_v<T>)
  void write(T value) {
    using Raw = WireRawTypeT<T>;
    if constexpr (std::is_floating_point_v<T>) {
      static_assert(sizeof(T) == sizeof(std::uint32_t));
      write(std::bit_cast<std::uint32_t>(value));
    } else {
      using Unsigned = std::make_unsigned_t<Raw>;
      const auto raw = static_cast<Unsigned>(static_cast<Raw>(value));
      for (std::size_t i = 0; i < sizeof(Unsigned); ++i) {
        bytes_.push_back(static_cast<std::byte>((raw >> (i * 8U)) & 0xffU));
      }
    }
  }

  void write_bytes(std::span<const std::byte> bytes) {
    bytes_.insert(bytes_.end(), bytes.begin(), bytes.end());
  }

  [[nodiscard]] const std::vector<std::byte>& bytes() const { return bytes_; }
  [[nodiscard]] std::vector<std::byte> take() { return std::move(bytes_); }

 private:
  std::vector<std::byte> bytes_;
};

class Reader {
 public:
  explicit Reader(std::span<const std::byte> bytes) : bytes_(bytes) {}

  template <typename T>
    requires(std::is_integral_v<T> || std::is_enum_v<T> || std::is_floating_point_v<T>)
  [[nodiscard]] T read() {
    using Raw = WireRawTypeT<T>;
    if constexpr (std::is_floating_point_v<T>) {
      static_assert(sizeof(T) == sizeof(std::uint32_t));
      return std::bit_cast<T>(read<std::uint32_t>());
    } else {
      using Unsigned = std::make_unsigned_t<Raw>;
      require(sizeof(Unsigned));
      Unsigned value{};
      for (std::size_t i = 0; i < sizeof(Unsigned); ++i) {
        value |= static_cast<Unsigned>(std::to_integer<unsigned int>(bytes_[offset_ + i]))
                 << (i * 8U);
      }
      offset_ += sizeof(Unsigned);
      return static_cast<T>(static_cast<Raw>(value));
    }
  }

  [[nodiscard]] std::vector<std::byte> read_bytes(std::size_t size) {
    require(size);
    const auto begin = bytes_.begin() + static_cast<std::ptrdiff_t>(offset_);
    const auto end = begin + static_cast<std::ptrdiff_t>(size);
    offset_ += size;
    return {begin, end};
  }

  [[nodiscard]] std::size_t remaining() const { return bytes_.size() - offset_; }

 private:
  void require(std::size_t size) const {
    if (size > remaining()) {
      throw ProtocolError("message ended before the declared payload was complete");
    }
  }

  std::span<const std::byte> bytes_;
  std::size_t offset_{};
};

void write_pose(Writer& writer, const Pose& pose) {
  for (const float component : pose.position) {
    writer.write(component);
  }
  for (const float component : pose.orientation) {
    writer.write(component);
  }
  writer.write(pose.flags);
}

[[nodiscard]] Pose read_pose(Reader& reader) {
  Pose pose;
  for (float& component : pose.position) {
    component = reader.read<float>();
  }
  for (float& component : pose.orientation) {
    component = reader.read<float>();
  }
  pose.flags = reader.read<std::uint32_t>();
  return pose;
}

void write_controller(Writer& writer, const ControllerState& controller) {
  write_pose(writer, controller.pose);
  writer.write(controller.buttons);
  for (const float value : controller.thumbstick) {
    writer.write(value);
  }
  writer.write(controller.trigger);
  writer.write(controller.grip);
}

[[nodiscard]] ControllerState read_controller(Reader& reader) {
  ControllerState controller;
  controller.pose = read_pose(reader);
  controller.buttons = reader.read<std::uint64_t>();
  for (float& value : controller.thumbstick) {
    value = reader.read<float>();
  }
  controller.trigger = reader.read<float>();
  controller.grip = reader.read<float>();
  return controller;
}

void write_eye_view(Writer& writer, const EyeView& view) {
  write_pose(writer, view.pose);
  writer.write(view.fov.angle_left);
  writer.write(view.fov.angle_right);
  writer.write(view.fov.angle_up);
  writer.write(view.fov.angle_down);
}

void write_hand_joint(Writer& writer, const HandJoint& joint) {
  write_pose(writer, joint.pose);
  writer.write(joint.radius);
}

[[nodiscard]] HandJoint read_hand_joint(Reader& reader) {
  return {.pose = read_pose(reader), .radius = reader.read<float>()};
}

[[nodiscard]] EyeView read_eye_view(Reader& reader) {
  EyeView view;
  view.pose = read_pose(reader);
  view.fov.angle_left = reader.read<float>();
  view.fov.angle_right = reader.read<float>();
  view.fov.angle_up = reader.read<float>();
  view.fov.angle_down = reader.read<float>();
  return view;
}

[[nodiscard]] MessageType type_of(const Payload& payload) {
  return std::visit(
      [](const auto& value) -> MessageType {
        using T = std::decay_t<decltype(value)>;
        if constexpr (std::is_same_v<T, VideoFrame>) {
          return MessageType::VideoFrame;
        } else if constexpr (std::is_same_v<T, PoseInput>) {
          return MessageType::PoseInput;
        } else if constexpr (std::is_same_v<T, ControlMessage>) {
          return MessageType::Control;
        } else if constexpr (std::is_same_v<T, HapticCommand>) {
          return MessageType::HapticCommand;
        } else {
          return MessageType::HandTrackingInput;
        }
      },
      payload);
}

[[nodiscard]] std::vector<std::byte> serialize_payload(const Payload& payload) {
  Writer writer;
  std::visit(
      [&writer](const auto& value) {
        using T = std::decay_t<decltype(value)>;
        if constexpr (std::is_same_v<T, VideoFrame>) {
          writer.write(value.capture_timestamp_ns);
          for (const auto& view : value.render_views) {
            write_eye_view(writer, view);
          }
          writer.write(value.width);
          writer.write(value.height);
          writer.write(value.codec);
          writer.write(value.eye_count);
          writer.write(value.flags);
          if (value.encoded_data.size() > kMaxPayloadSize) {
            throw ProtocolError("video frame is larger than the protocol payload limit");
          }
          writer.write(static_cast<std::uint32_t>(value.encoded_data.size()));
          writer.write_bytes(value.encoded_data);
        } else if constexpr (std::is_same_v<T, PoseInput>) {
          writer.write(value.sample_timestamp_ns);
          write_pose(writer, value.head);
          write_controller(writer, value.left);
          write_controller(writer, value.right);
        } else if constexpr (std::is_same_v<T, ControlMessage>) {
          writer.write(value.kind);
          writer.write(value.flags);
          writer.write(value.timestamp_ns);
          if (value.data.size() > kMaxPayloadSize) {
            throw ProtocolError("control data is larger than the protocol payload limit");
          }
          writer.write(static_cast<std::uint32_t>(value.data.size()));
          writer.write_bytes(value.data);
        } else if constexpr (std::is_same_v<T, HapticCommand>) {
          writer.write(value.timestamp_ns);
          writer.write(value.side);
          writer.write(value.action);
          writer.write(value.reserved);
          writer.write(value.amplitude);
          writer.write(value.frequency_hz);
          writer.write(value.duration_ns);
        } else {
          writer.write(value.sample_timestamp_ns);
          writer.write(static_cast<std::uint8_t>(value.left_active));
          writer.write(static_cast<std::uint8_t>(value.right_active));
          writer.write(static_cast<std::uint16_t>(kHandJointCount));
          for (const auto& joint : value.left_joints) write_hand_joint(writer, joint);
          for (const auto& joint : value.right_joints) write_hand_joint(writer, joint);
        }
      },
      payload);
  return writer.take();
}

void require_no_trailing_bytes(const Reader& reader) {
  if (reader.remaining() != 0) {
    throw ProtocolError("payload contains trailing bytes");
  }
}

[[nodiscard]] Payload deserialize_payload(MessageType type, std::span<const std::byte> bytes) {
  Reader reader(bytes);
  switch (type) {
    case MessageType::VideoFrame: {
      VideoFrame value;
      value.capture_timestamp_ns = reader.read<std::uint64_t>();
      for (auto& view : value.render_views) {
        view = read_eye_view(reader);
      }
      value.width = reader.read<std::uint32_t>();
      value.height = reader.read<std::uint32_t>();
      value.codec = reader.read<VideoCodec>();
      value.eye_count = reader.read<std::uint8_t>();
      value.flags = reader.read<std::uint16_t>();
      const auto data_size = reader.read<std::uint32_t>();
      value.encoded_data = reader.read_bytes(data_size);
      require_no_trailing_bytes(reader);
      return value;
    }
    case MessageType::PoseInput: {
      PoseInput value;
      value.sample_timestamp_ns = reader.read<std::uint64_t>();
      value.head = read_pose(reader);
      value.left = read_controller(reader);
      value.right = read_controller(reader);
      require_no_trailing_bytes(reader);
      return value;
    }
    case MessageType::Control: {
      ControlMessage value;
      value.kind = reader.read<ControlKind>();
      value.flags = reader.read<std::uint16_t>();
      value.timestamp_ns = reader.read<std::uint64_t>();
      const auto data_size = reader.read<std::uint32_t>();
      value.data = reader.read_bytes(data_size);
      require_no_trailing_bytes(reader);
      return value;
    }
    case MessageType::HapticCommand: {
      HapticCommand value;
      value.timestamp_ns = reader.read<std::uint64_t>();
      value.side = reader.read<HandSide>();
      value.action = reader.read<HapticAction>();
      value.reserved = reader.read<std::uint16_t>();
      value.amplitude = reader.read<float>();
      value.frequency_hz = reader.read<float>();
      value.duration_ns = reader.read<std::uint64_t>();
      if ((value.side != HandSide::Left && value.side != HandSide::Right) ||
          (value.action != HapticAction::Apply && value.action != HapticAction::Stop)) {
        throw ProtocolError("invalid haptic command enum value");
      }
      require_no_trailing_bytes(reader);
      return value;
    }
    case MessageType::HandTrackingInput: {
      HandTrackingInput value;
      value.sample_timestamp_ns = reader.read<std::uint64_t>();
      value.left_active = reader.read<std::uint8_t>() != 0;
      value.right_active = reader.read<std::uint8_t>() != 0;
      const auto joint_count = reader.read<std::uint16_t>();
      if (joint_count != kHandJointCount) {
        throw ProtocolError("hand tracking joint count is not 26");
      }
      for (auto& joint : value.left_joints) joint = read_hand_joint(reader);
      for (auto& joint : value.right_joints) joint = read_hand_joint(reader);
      require_no_trailing_bytes(reader);
      return value;
    }
  }
  throw ProtocolError("unknown message type");
}

}  // namespace

MessageHeader parse_header(std::span<const std::byte> bytes) {
  if (bytes.size() < kHeaderSize) {
    throw ProtocolError("message is shorter than the protocol header");
  }
  Reader reader(bytes.first(kHeaderSize));
  MessageHeader header;
  header.magic = reader.read<std::uint32_t>();
  header.version = reader.read<std::uint16_t>();
  header.type = reader.read<MessageType>();
  header.payload_size = reader.read<std::uint32_t>();
  header.sequence = reader.read<std::uint64_t>();
  if (header.magic != kMagic) {
    throw ProtocolError("invalid protocol magic");
  }
  if (header.version != kProtocolVersion) {
    throw ProtocolError("unsupported protocol version");
  }
  if (header.payload_size > kMaxPayloadSize) {
    throw ProtocolError("declared payload exceeds the protocol limit");
  }
  switch (header.type) {
    case MessageType::VideoFrame:
    case MessageType::PoseInput:
    case MessageType::Control:
    case MessageType::HapticCommand:
    case MessageType::HandTrackingInput:
      break;
    default:
      throw ProtocolError("unknown message type");
  }
  return header;
}

std::vector<std::byte> serialize(const Message& message) {
  auto payload = serialize_payload(message.payload);
  if (payload.size() > kMaxPayloadSize ||
      payload.size() > std::numeric_limits<std::uint32_t>::max()) {
    throw ProtocolError("serialized payload exceeds the protocol limit");
  }

  Writer writer;
  writer.write(kMagic);
  writer.write(kProtocolVersion);
  writer.write(type_of(message.payload));
  writer.write(static_cast<std::uint32_t>(payload.size()));
  writer.write(message.sequence);
  writer.write_bytes(payload);
  return writer.take();
}

Message deserialize(std::span<const std::byte> bytes) {
  const auto header = parse_header(bytes);
  const auto total_size = kHeaderSize + static_cast<std::size_t>(header.payload_size);
  if (bytes.size() != total_size) {
    throw ProtocolError(bytes.size() < total_size ? "message is truncated"
                                                   : "message contains trailing bytes");
  }
  return Message{
      .sequence = header.sequence,
      .payload = deserialize_payload(header.type, bytes.subspan(kHeaderSize)),
  };
}

}  // namespace maquestlink::protocol

File: quest-client/Assets/MetalQuestLink/Runtime/Protocol.cs
using System;
using System.IO;
using System.Text;

namespace MetalQuestLink.QuestClient
{
    public enum MessageType : ushort
    {
        VideoFrame = 1,
        PoseInput = 2,
        Control = 3,
        HapticCommand = 4,
        HandTrackingInput = 5,
    }

    public enum VideoCodec : byte
    {
        H264 = 1,
        Hevc = 2,
    }

    [Flags]
    public enum VideoFrameFlags : ushort
    {
        KeyFrame = 1 << 0,
        Passthrough = 1 << 1,
        ChromaKeyTransparency = 1 << 2,
    }

    public enum HandSide : byte
    {
        Left = 1,
        Right = 2,
    }

    public enum HapticAction : byte
    {
        Apply = 1,
        Stop = 2,
    }

    public enum ControlKind : ushort
    {
        Hello = 1,
        HelloAck = 2,
        StartStream = 3,
        StopStream = 4,
        Ping = 5,
        Pong = 6,
        Disconnect = 7,
    }

    [Flags]
    public enum PoseFlags : uint
    {
        PositionValid = 1u << 0,
        OrientationValid = 1u << 1,
        PositionTracked = 1u << 2,
        OrientationTracked = 1u << 3,
    }

    [Flags]
    public enum ControllerButtons : ulong
    {
        PrimaryButton = 1ul << 0,
        SecondaryButton = 1ul << 1,
        ThumbstickButton = 1ul << 2,
        MenuButton = 1ul << 3,
        PrimaryTouch = 1ul << 4,
        SecondaryTouch = 1ul << 5,
        ThumbstickTouch = 1ul << 6,
        TriggerTouch = 1ul << 7,
    }

    public struct Vector2f
    {
        public float X;
        public float Y;
    }

    public struct Vector3f
    {
        public float X;
        public float Y;
        public float Z;
    }

    public struct Quaternionf
    {
        public float X;
        public float Y;
        public float Z;
        public float W;
    }

    public struct PoseState
    {
        public Vector3f Position;
        public Quaternionf Orientation;
        public PoseFlags Flags;
    }

    public struct ControllerState
    {
        public PoseState Pose;
        public ControllerButtons Buttons;
        public Vector2f Thumbstick;
        public float Trigger;
        public float Grip;
    }

    public struct FieldOfView
    {
        public float AngleLeft;
        public float AngleRight;
        public float AngleUp;
        public float AngleDown;
    }

    public struct EyeView
    {
        public PoseState Pose;
        public FieldOfView Fov;
    }

    public sealed class VideoFrame
    {
        public ulong CaptureTimestampNs;
        public EyeView[] RenderViews = new EyeView[2];
        public uint Width;
        public uint Height;
        public VideoCodec Codec = VideoCodec.H264;
        public byte EyeCount = 2;
        public ushort Flags;
        public byte[] EncodedData = Array.Empty<byte>();
        // Quest-local receive time. This is transport metadata and is not serialized.
        public ulong ReceiveTimestampNs;
    }

    public sealed class PoseInput
    {
        public ulong SampleTimestampNs;
        public PoseState Head;
        public ControllerState Left;
        public ControllerState Right;
    }

    public sealed class ControlMessage
    {
        public ControlKind Kind;
        public ushort Flags;
        public ulong TimestampNs;
        public byte[] Data = Array.Empty<byte>();
    }

    public sealed class HapticCommand
    {
        public ulong TimestampNs;
        public HandSide Side;
        public HapticAction Action;
        public ushort Reserved;
        public float Amplitude;
        public float FrequencyHz;
        public ulong DurationNs;
    }

    public struct HandJointState
    {
        public PoseState Pose;
        public float Radius;
    }

    public sealed class HandTrackingInput
    {
        public ulong SampleTimestampNs;
        public bool LeftActive;
        public bool RightActive;
        public HandJointState[] LeftJoints = new HandJointState[Protocol.HandJointCount];
        public HandJointState[] RightJoints = new HandJointState[Protocol.HandJointCount];
    }

    public sealed class WireMessage
    {
        public ulong Sequence;
        public object Payload;

        public WireMessage(ulong sequence, object payload)
        {
            Sequence = sequence;
            Payload = payload ?? throw new ArgumentNullException(nameof(payload));
        }
    }

    public struct MessageHeader
    {
        public uint Magic;
        public ushort Version;
        public MessageType Type;
        public uint PayloadSize;
        public ulong Sequence;
    }

    public sealed class ProtocolException : Exception
    {
        public ProtocolException(string message) : base(message) { }
    }

    public static class Protocol
    {
        public const uint Magic = 0x4b4c544d;
        public const ushort Version = 1;
        public const int HeaderSize = 20;
        public const float PassthroughApproximationAlpha = 0.82f;
        public const uint MaxPayloadSize = 64u * 1024u * 1024u;
        public const int HandJointCount = 26;

        public static MessageHeader ParseHeader(byte[] bytes)
        {
            if (bytes == null || bytes.Length < HeaderSize)
            {
                throw new ProtocolException("message is shorter than the protocol header");
            }

            var reader = new WireReader(bytes, 0, HeaderSize);
            var header = new MessageHeader
            {
                Magic = reader.ReadUInt32(),
                Version = reader.ReadUInt16(),
                Type = (MessageType)reader.ReadUInt16(),
                PayloadSize = reader.ReadUInt32(),
                Sequence = reader.ReadUInt64(),
            };
            if (header.Magic != Magic)
            {
                throw new ProtocolException("invalid protocol magic");
            }
            if (header.Version != Version)
            {
                throw new ProtocolException("unsupported protocol version");
            }
            if (header.PayloadSize > MaxPayloadSize)
            {
                throw new ProtocolException("declared payload exceeds the protocol limit");
            }
            if (header.Type != MessageType.VideoFrame && header.Type != MessageType.PoseInput &&
                header.Type != MessageType.Control && header.Type != MessageType.HapticCommand &&
                header.Type != MessageType.HandTrackingInput)
            {
                throw new ProtocolException("unknown message type");
            }
            return header;
        }

        public static byte[] Serialize(WireMessage message)
        {
            if (message == null)
            {
                throw new ArgumentNullException(nameof(message));
            }

            MessageType type;
            byte[] payload;
            if (message.Payload is VideoFrame video)
            {
                type = MessageType.VideoFrame;
                payload = SerializeVideo(video);
            }
            else if (message.Payload is PoseInput pose)
            {
                type = MessageType.PoseInput;
                payload = SerializePoseInput(pose);
            }
            else if (message.Payload is ControlMessage control)
            {
                type = MessageType.Control;
                payload = SerializeControl(control);
            }
            else if (message.Payload is HapticCommand haptic)
            {
                type = MessageType.HapticCommand;
                payload = SerializeHaptic(haptic);
            }
            else if (message.Payload is HandTrackingInput hands)
            {
                type = MessageType.HandTrackingInput;
                payload = SerializeHands(hands);
            }
            else
            {
                throw new ProtocolException("unsupported message payload type");
            }

            if ((uint)payload.Length > MaxPayloadSize)
            {
                throw new ProtocolException("serialized payload exceeds the protocol limit");
            }

            var writer = new WireWriter(HeaderSize + payload.Length);
            writer.Write(Magic);
            writer.Write(Version);
            writer.Write((ushort)type);
            writer.Write((uint)payload.Length);
            writer.Write(message.Sequence);
            writer.Write(payload);
            return writer.ToArray();
        }

        public static WireMessage Deserialize(byte[] bytes)
        {
            var header = ParseHeader(bytes);
            var expectedSize = HeaderSize + checked((int)header.PayloadSize);
            if (bytes.Length != expectedSize)
            {
                throw new ProtocolException(bytes.Length < expectedSize
                    ? "message is truncated"
                    : "message contains trailing bytes");
            }

            var reader = new WireReader(bytes, HeaderSize, (int)header.PayloadSize);
            object payload;
            switch (header.Type)
            {
                case MessageType.VideoFrame:
                    payload = DeserializeVideo(reader);
                    break;
                case MessageType.PoseInput:
                    payload = DeserializePoseInput(reader);
                    break;
                case MessageType.Control:
                    payload = DeserializeControl(reader);
                    break;
                case MessageType.HapticCommand:
                    payload = DeserializeHaptic(reader);
                    break;
                case MessageType.HandTrackingInput:
                    payload = DeserializeHands(reader);
                    break;
                default:
                    throw new ProtocolException("unknown message type");
            }
            reader.RequireComplete();
            return new WireMessage(header.Sequence, payload);
        }

        private static byte[] SerializeVideo(VideoFrame value)
        {
            if (value.RenderViews == null || value.RenderViews.Length != 2)
            {
                throw new ProtocolException("video frame must contain exactly two render views");
            }
            var data = value.EncodedData ?? Array.Empty<byte>();
            var writer = new WireWriter(120 + data.Length);
            writer.Write(value.CaptureTimestampNs);
            WriteEyeView(writer, value.RenderViews[0]);
            WriteEyeView(writer, value.RenderViews[1]);
            writer.Write(value.Width);
            writer.Write(value.Height);
            writer.Write((byte)value.Codec);
            writer.Write(value.EyeCount);
            writer.Write(value.Flags);
            writer.Write((uint)data.Length);
            writer.Write(data);
            return writer.ToArray();
        }

        private static byte[] SerializePoseInput(PoseInput value)
        {
            var writer = new WireWriter(152);
            writer.Write(value.SampleTimestampNs);
            WritePose(writer, value.Head);
            WriteController(writer, value.Left);
            WriteController(writer, value.Right);
            return writer.ToArray();
        }

        private static byte[] SerializeControl(ControlMessage value)
        {
            var data = value.Data ?? Array.Empty<byte>();
            var writer = new WireWriter(16 + data.Length);
            writer.Write((ushort)value.Kind);
            writer.Write(value.Flags);
            writer.Write(value.TimestampNs);
            writer.Write((uint)data.Length);
            writer.Write(data);
            return writer.ToArray();
        }

        private static byte[] SerializeHaptic(HapticCommand value)
        {
            var writer = new WireWriter(28);
            writer.Write(value.TimestampNs);
            writer.Write((byte)value.Side);
            writer.Write((byte)value.Action);
            writer.Write(value.Reserved);
            writer.Write(value.Amplitude);
            writer.Write(value.FrequencyHz);
            writer.Write(value.DurationNs);
            return writer.ToArray();
        }

        private static byte[] SerializeHands(HandTrackingInput value)
        {
            if (value.LeftJoints == null || value.LeftJoints.Length != HandJointCount ||
                value.RightJoints == null || value.RightJoints.Length != HandJointCount)
            {
                throw new ProtocolException("hand tracking must contain 26 joints per hand");
            }
            var writer = new WireWriter(12 + HandJointCount * 2 * 36);
            writer.Write(value.SampleTimestampNs);
            writer.Write((byte)(value.LeftActive ? 1 : 0));
            writer.Write((byte)(value.RightActive ? 1 : 0));
            writer.Write((ushort)HandJointCount);
            foreach (var joint in value.LeftJoints) WriteHandJoint(writer, joint);
            foreach (var joint in value.RightJoints) WriteHandJoint(writer, joint);
            return writer.ToArray();
        }

        private static VideoFrame DeserializeVideo(WireReader reader)
        {
            var value = new VideoFrame { CaptureTimestampNs = reader.ReadUInt64() };
            value.RenderViews[0] = ReadEyeView(reader);
            value.RenderViews[1] = ReadEyeView(reader);
            value.Width = reader.ReadUInt32();
            value.Height = reader.ReadUInt32();
            value.Codec = (VideoCodec)reader.ReadByte();
            value.EyeCount = reader.ReadByte();
            value.Flags = reader.ReadUInt16();
            value.EncodedData = reader.ReadBytes(checked((int)reader.ReadUInt32()));
            return value;
        }

        private static PoseInput DeserializePoseInput(WireReader reader)
        {
            return new PoseInput
            {
                SampleTimestampNs = reader.ReadUInt64(),
                Head = ReadPose(reader),
                Left = ReadController(reader),
                Right = ReadController(reader),
            };
        }

        private static ControlMessage DeserializeControl(WireReader reader)
        {
            var value = new ControlMessage
            {
                Kind = (ControlKind)reader.ReadUInt16(),
                Flags = reader.ReadUInt16(),
                TimestampNs = reader.ReadUInt64(),
            };
            value.Data = reader.ReadBytes(checked((int)reader.ReadUInt32()));
            return value;
        }

        private static HapticCommand DeserializeHaptic(WireReader reader)
        {
            var value = new HapticCommand
            {
                TimestampNs = reader.ReadUInt64(),
                Side = (HandSide)reader.ReadByte(),
                Action = (HapticAction)reader.ReadByte(),
                Reserved = reader.ReadUInt16(),
                Amplitude = reader.ReadSingle(),
                FrequencyHz = reader.ReadSingle(),
                DurationNs = reader.ReadUInt64(),
            };
            if ((value.Side != HandSide.Left && value.Side != HandSide.Right) ||
                (value.Action != HapticAction.Apply && value.Action != HapticAction.Stop))
            {
                throw new ProtocolException("invalid haptic command enum value");
            }
            return value;
        }

        private static HandTrackingInput DeserializeHands(WireReader reader)
        {
            var value = new HandTrackingInput
            {
                SampleTimestampNs = reader.ReadUInt64(),
                LeftActive = reader.ReadByte() != 0,
                RightActive = reader.ReadByte() != 0,
            };
            if (reader.ReadUInt16() != HandJointCount)
            {
                throw new ProtocolException("hand tracking joint count is not 26");
            }
            for (var index = 0; index < HandJointCount; index++) value.LeftJoints[index] = ReadHandJoint(reader);
            for (var index = 0; index < HandJointCount; index++) value.RightJoints[index] = ReadHandJoint(reader);
            return value;
        }

        private static void WritePose(WireWriter writer, PoseState value)
        {
            writer.Write(value.Position.X);
            writer.Write(value.Position.Y);
            writer.Write(value.Position.Z);
            writer.Write(value.Orientation.X);
            writer.Write(value.Orientation.Y);
            writer.Write(value.Orientation.Z);
            writer.Write(value.Orientation.W);
            writer.Write((uint)value.Flags);
        }

        private static PoseState ReadPose(WireReader reader)
        {
            return new PoseState
            {
                Position = new Vector3f { X = reader.ReadSingle(), Y = reader.ReadSingle(), Z = reader.ReadSingle() },
                Orientation = new Quaternionf
                {
                    X = reader.ReadSingle(), Y = reader.ReadSingle(), Z = reader.ReadSingle(), W = reader.ReadSingle(),
                },
                Flags = (PoseFlags)reader.ReadUInt32(),
            };
        }

        private static void WriteHandJoint(WireWriter writer, HandJointState value)
        {
            WritePose(writer, value.Pose);
            writer.Write(value.Radius);
        }

        private static HandJointState ReadHandJoint(WireReader reader)
        {
            return new HandJointState { Pose = ReadPose(reader), Radius = reader.ReadSingle() };
        }

        private static void WriteController(WireWriter writer, ControllerState value)
        {
            WritePose(writer, value.Pose);
            writer.Write((ulong)value.Buttons);
            writer.Write(value.Thumbstick.X);
            writer.Write(value.Thumbstick.Y);
            writer.Write(value.Trigger);
            writer.Write(value.Grip);
        }

        private static ControllerState ReadController(WireReader reader)
        {
            return new ControllerState
            {
                Pose = ReadPose(reader),
                Buttons = (ControllerButtons)reader.ReadUInt64(),
                Thumbstick = new Vector2f { X = reader.ReadSingle(), Y = reader.ReadSingle() },
                Trigger = reader.ReadSingle(),
                Grip = reader.ReadSingle(),
            };
        }

        private static void WriteEyeView(WireWriter writer, EyeView value)
        {
            WritePose(writer, value.Pose);
            writer.Write(value.Fov.AngleLeft);
            writer.Write(value.Fov.AngleRight);
            writer.Write(value.Fov.AngleUp);
            writer.Write(value.Fov.AngleDown);
        }

        private static EyeView ReadEyeView(WireReader reader)
        {
            return new EyeView
            {
                Pose = ReadPose(reader),
                Fov = new FieldOfView
                {
                    AngleLeft = reader.ReadSingle(),
                    AngleRight = reader.ReadSingle(),
                    AngleUp = reader.ReadSingle(),
                    AngleDown = reader.ReadSingle(),
                },
            };
        }

        private sealed class WireWriter
        {
            private readonly MemoryStream stream;
            private readonly BinaryWriter writer;

            public WireWriter(int capacity)
            {
                stream = new MemoryStream(capacity);
                writer = new BinaryWriter(stream, Encoding.UTF8, true);
            }

            public void Write(byte value) => writer.Write(value);
            public void Write(ushort value) => writer.Write(value);
            public void Write(uint value) => writer.Write(value);
            public void Write(ulong value) => writer.Write(value);
            public void Write(float value) => writer.Write(value);
            public void Write(byte[] value) => writer.Write(value);
            public byte[] ToArray() => stream.ToArray();
        }

        private sealed class WireReader
        {
            private readonly MemoryStream stream;
            private readonly BinaryReader reader;
            private readonly long end;

            public WireReader(byte[] bytes, int offset, int length)
            {
                stream = new MemoryStream(bytes, offset, length, false);
                reader = new BinaryReader(stream, Encoding.UTF8, true);
                // MemoryStream(byte[], index, count) exposes Position relative to index.
                end = length;
            }

            public byte ReadByte() { Require(1); return reader.ReadByte(); }
            public ushort ReadUInt16() { Require(2); return reader.ReadUInt16(); }
            public uint ReadUInt32() { Require(4); return reader.ReadUInt32(); }
            public ulong ReadUInt64() { Require(8); return reader.ReadUInt64(); }
            public float ReadSingle() { Require(4); return reader.ReadSingle(); }

            public byte[] ReadBytes(int count)
            {
                Require(count);
                var result = reader.ReadBytes(count);
                if (result.Length != count)
                {
                    throw new ProtocolException("message ended before the declared payload was complete");
                }
                return result;
            }

            public void RequireComplete()
            {
                if (stream.Position != end)
                {
                    throw new ProtocolException("payload contains trailing bytes");
                }
            }

            private void Require(int count)
            {
                if (count < 0 || stream.Position + count > end)
                {
                    throw new ProtocolException("message ended before the declared payload was complete");
                }
            }
        }
    }
}
